### DISCUSSION TITLE
Cypress selector issue data attribute vs element ID

### DISCUSSION THREAD

**Discussion by rmnrdi** [cypress-io/cypress#27456] [Questions and Help]
I'm trying to get an element by data attribute data-el="wmMainMenu". Cypress fails and **can't** find it because it considers the element hidden (due to an attribute). 
https://docs.cypress.io/guides/core-concepts/interacting-with-elements#Visibility

If I get the element by id #TfrmHome_wcMenu3, Cypress **can** find the element.

Also, if this is true, then this is BS because data-* doesn't always work.

From Cypress docs
**"Oftentimes we see users run into problems targeting their elements because:**

Your application may use dynamic classes or ID's that change
**Your selectors break from development changes to CSS styles or JS behavior**

Luckily, it is possible to avoid both of these problems.

Don't target elements based on CSS attributes such as: id, class, tag
Don't target elements that may change their textContent
**Add data-* attributes to make it easier to target elements"** 

Doesn't seem to work with custom data attributes like ours
 **data-el="ComponentName"** 

I got a response from ChatGPT saying Cypress treats different selectors differently. 
If this is so, can someone explain why and what is the solution to this issue? 

Also, if this is true, then this chart is wrong because data-* doesn't always work.
![image](https://github.com/cypress-io/cypress/assets/23411447/5c28869f-53a5-4ab3-ac91-311f0be2a68f)

**Comment by portyu9** [comment]
Hi @rmnrdi 

No—Cypress does **not** treat an ID selector as more visible or more actionable than a `data-*` selector.

These:

```js
cy.get('[data-el="wmMainMenu"]')
cy.get('#TfrmHome_wcMenu3')
```

are simply two CSS queries. Once Cypress has found the element, the same visibility/actionability rules apply.

If one works and the other reports a hidden element, the selectors are almost certainly resolving to **different DOM elements**, or the data selector is matching multiple elements and at least one of them is hidden.

I would inspect that directly:

```js
cy.get('[data-el="wmMainMenu"]')
  .should('have.length', 1)
  .should('be.visible')
```

and compare the actual element with:

```js
cy.get('#TfrmHome_wcMenu3')
  .should('be.visible')
```

Check especially for duplicate responsive/mobile menus, hidden ancestors, or the `data-el` attribute being placed on a wrapper while the ID is on the interactive child.

In fact, dedicated `data-cy`/`data-*` attributes are generally **recommended** by Cypress because they make selectors more stable. That recommendation is about selector stability, not visibility.

Please mark this answer as accepted if it helped, thank you.
